lsat/store.go

```diff
@@ -63,33 +69,76 @@ func (f *FileStore) HasToken() bool {
 	return fileExists(f.fileName)
 }
 
-// Token returns the token that is contained in the store or an error if there
-// is none.
+// CurrentToken returns the token that is currently contained in the store or an
+// error if there is none.
 //
 // NOTE: This is part of the Store interface.
-func (f *FileStore) Token() (*Token, error) {
+func (f *FileStore) CurrentToken() (*Token, error) {
 	if !f.HasToken() {
 		return nil, ErrNoToken
 	}
-	bytes, err := ioutil.ReadFile(f.fileName)
+	return readTokenFile(f.fileName)
+}
+
+// AllTokens returns all tokens that the store has knowledge of, even
+// if they might be expired.
+func (f *FileStore) AllTokens() ([]*Token, error) {
+	tokens := make([]*Token, 0)
+
+	// All tokens start with the same name but old tokens have a timestamp
+	// appended to the original file name.
+	tokenDir := filepath.Dir(f.fileName)
+	files, err := ioutil.ReadDir(tokenDir)
 	if err != nil {
 		return nil, err
 	}
-	return deserializeToken(bytes)
+	for _, file := range files {
+		name := file.Name()
+		if !strings.HasPrefix(name, storeFileName) {
+			continue
+		}
+		token, err := readTokenFile(filepath.Join(tokenDir, name))
+		if err != nil {
+			return nil, err
+		}
+		tokens = append(tokens, token)
+	}
+
+	return tokens, nil
 }
 
 // StoreToken saves a token to the store, overwriting any old token if there is
 // one.
 //
 // NOTE: This is part of the Store interface.
 func (f *FileStore) StoreToken(token *Token) error {
+	// Don't just overwrite existing old tokens but create a backup for
+	// accounting purposes.
+	if f.HasToken() {
+		timestampSuffix := time.Now().Format("2006-01-02T15-04-05")
+		err := os.Rename(f.fileName, f.fileName+timestampSuffix)
```

Comment:
We'll want to make sure this rename along with the new token write below are atomic.

Reply:
On windows, we also need to _close_ the file (if it's already open) before we rename it. We've hit this a few times before in neutrino and also the way we handle SCB files on disk.

Reply:
Ok, I tried making this as atomic as possible.